### PR TITLE
WIP/suggestion: Add DI framework and rewrite application to use this pattern

### PR DESCRIPTION
**Pull request by pbjorklund**
Logger was reliant on microsoft ajax (String.format). Refactored this to use typescript string itnerpolation.
Added beginning for tests for objectpropertybagentries. Right now it just tests that it returns a promise when ran inside a ctx from Util.

Questions: The import pattern used here. Is this the way to do it?

Note: The ```/// <reference``` must come before the 'use strict'; for it to work.

### CODE REVIEW

**Comment by pbjorklund** [comment]
Logging should probably be done with decorators and not by coupling logging to a base class (or base class should at least be made abstract).  We can see why in this PR where I needed up to update an unrelated and optional piece of code to get the class under test compiling. 

Why do we extend ObjectHandlerBase instead of having handlers implement the IObjectHandler interface? If we did that we could easily create an LoggingObjectListsDecorator that implements IObjectHandler and decorated another IObjectHandler.

**Comment by patrick-rodgers** [comment]
Good points - let me pull in @olemp who wrote the provisioning pieces. We do have an abstract logger at the root which this code should be updated to use. It was originally ported in from another project.

**Comment by olemp** [comment]
Good points @pbjorklund 

This is the reason why I decided to merge https://github.com/Puzzlepart/js-sites-core in here - input from as many people as possible.

We should probably use the same logger as the rest of the PnP. No reason as far as I can see having a separate logger for the provisioning part.

**Why do we extend ObjectHandlerBase instead of having handlers implement the IObjectHandler interface?**
Good question. I just thought this was common practice. If there's a better way to do this, we should change it.

**Comment by pbjorklund** [comment]
I added some thoughts on DI to the yammer thread.

https://www.yammer.com/itpronetwork/#/Threads/show?threadId=679052260

If that was used then we would resolve IObjectHandler to for instance LoggingObjectHandlerDecorator (Or use the TS decorator's which I'm not super confident on how they work yet)

**Comment by patrick-rodgers** [comment]
Help me out on this one, where are we at - should this be merged as-is or do we want to close it and submit a new one with just parts of this? I read back through the discussion and I am not sure if all the questions are resolved or where we are coming down.

**Comment by pbjorklund** [comment]
I'm evaluating http://inversify.io/ for DI right now, it's taking longer then I thought so no real conclusions yet.

The actual code in the pr could be merged I guess since it's not rally related

**Comment by olemp** [comment]
@patrick-rodgers @pbjorklund I've been out of the loop for a while now.. Hope to get back in soon :)

I put my trust in @pbjorklund

**Comment by pbjorklund** [comment]
Well @olemp, right now I'm not so sure that trusting me is the right decision ;)

I'm regularly pushing my ... experiments around DI to:
https://github.com/pbjorklund/PnP-JS-Core/tree/ioc/src (it's rebased it to use pr #60)

In https://github.com/pbjorklund/PnP-JS-Core/blob/ioc/src/pnp.test.ts we are now fetching a concrete implementation of PnPInterface (yeah, turns out I'm actually not a 100% sure on what the philosophy behind interface naming in typescript with the I prefix is either) that uses DI to get it's Site and Rest implementations (I think, not having a debugger made it kinda hard to actually verify)

At first I was trying to build against only a SiteInterface, but due to the design of the fluent api that builds new concrete classes of itself inside for instance SIte that turned out to be quite hard/impossible. 

So, in order to continue my journey I created a SiteFactoryInterface&SiteFactory and started using constructor injection to take that interface and create the concrete type via infersify, now at least we can depend on an abstraction around Site&SiteInterface which would let us change the implementation of the concrete Site class however we want while still being able to build SiteInterface conforming types that requires constructor parameters. Being able to just inject whatever mocks we want in the constructor of system under test will make testing much easier imho.

Not really all that happy with the SiteInterface since it leaks through the inherited Queryable (inheritance in general is tricky here and perhaps it's my lack of TS skills that get's me confused). I would rather inject a QueryerInterface conforming object into things that wan't to be able to perform queries.

See https://github.com/pbjorklund/PnP-JS-Core/blob/ioc/src/sharepoint/rest/rest.ts and https://github.com/pbjorklund/PnP-JS-Core/blob/ioc/src/sharepoint/rest/site.ts

Or diff against upstream pr 60.

This is very very much a work in progress, but I like putting stuff there early enough so that we can decide if it's never going to make it into the repo due to the work needed.

Also, this probably shouldn't be done/discussed inside a PR that's almost unrelated.

**Comment by pbjorklund** [comment]
@patrick-rodgers would love some feedback on this. :)

**Comment by patrick-rodgers** [comment]
Reviewed this today - it seems like a lot of extra complexity for I am not sure what gain. For example, when would we want to create a different site? I guess the case you are looking at is the .web or .crossDomainWeb. To me having these different calls here isn't that big a deal and creates ultimately less complex code.

You can do something like:

function<T>factory(blah: { new (q: string | Queryable, path?: string): T }, q: string | Queryable, path?: string): T {
return new blah();
}

factory(Site, q, path);  // which will give you a new Site

SO if T was an interface and you passed in a class that implements that you could do something like what you are trying. But not sure it is worth all the work.

**Comment by pbjorklund** [comment]
The gain is that we can program against interfaces instead of concrete types. This helps a lot when writing unit/function tests. 

By using the kernel from infersify we get one place to specify the object graph of the application by specifying how interfaces are resolved at runtime (https://github.com/pbjorklund/PnP-JS-Core/blob/ioc/src/inversify.config.ts). You only specify that IThing resolves to CThing, when you need something that's built by constructor(thing1: IThing1, thing2: IThing2, thing3: IThing3, thing4: IThing4) the injector knows to to construct this object for us. See https://github.com/pbjorklund/PnP-JS-Core/blob/ioc/src/pnp.test.ts.

The example is limited since it doesn't really work well when we are only programming against interfaces in certain parts of the application. I had to create the factory to make it possible to inject anything at all into the fluid style since we can't, from what I know, ```new self```. I would preferably not let classes handle creating their dependencies.

One possible (but a bit harsh) thing we could do when conforming to the pattern is to add a tslint rule to forbid ```new```. And yes, this means that we don't manually create our concrete implementations anywhere. Not something I advocate, but it illustrates the point.

I personally really like using this pattern for building applications as it makes everything very easy to test. When we new up a PnP object now we get the whole tree of objects, but in a test all we care about is what the class does - not it's dependencies.

Take Queryable.
```
 public get(parser: (r: Response) => Promise<any> = this.defaultParser): Promise<any> {
        let client = new HttpClient();
        return client.get(this.toUrlAndQuery()) [......]
```
Wouldn't it be nice if our design let us inject a HttpClientInterface that the class relied on? Then we could create our test queryable like ```let sut = new Queryable(MockHttpClient)```.
(This being javascript we could expose the HttpClient as a property and rewrite it at runtime. But I have been bitten by strange things metaprogramming brings before, so I'm not a big fan).

For general info on DI https://en.wikipedia.org/wiki/Dependency_injection.

Anyone else have any thoughts on this? @mike-morrison @alex-randall ?

**Comment by pbjorklund** [comment]
Renamed the PR title, it was not descriptive.

**Comment by mike-morrison** [comment]
We can lose the "use strict"; from all of the files transpiled by TSC.  1.8 handles it for you.

>#### Modules are now emitted with a "use strict"; prologue

>Modules were always parsed in strict mode as per ES6, but for non-ES6 targets this was not respected in the generated code. Starting with TypeScript 1.8, emitted modules are always in strict mode. This shouldn’t have any visible changes in most code as TS considers most strict mode errors as errors at compile time, but it means that some things which used to silently fail at runtime in your TS code, like assigning to NaN, will now loudly fail. You can reference the MDN Article on strict mode for a detailed list of the differences between strict mode and non-strict mode.

My vote is for [Aurelia DI](http://aurelia.io/docs.html#/aurelia/dependency-injection/1.0.0-beta.1.2.3/doc/api/overview).  The intention all along for Aurelia was to make best in class pieces that could be used independently.  so they could all be used individually.  It plays very well with TS, in fact you can optionally autoinject when it is used with TypeScript.  Inversify might be the same.  I don't know much about it.  I trust Rob Eisenberg to deliver solid stuff.

It will be v1 soon, but have a look in [the repo](https://github.com/aurelia/dependency-injection)

No src changes in 2 months.  Look at branches, Perf work was finished 9 months ago.  They have released quite a few betas in the last few months and this code hasn't changed.  Less than 1000 lines of code with 2 small "Aurelia-" dependencies.  Everything in this library conforms to ES2015.

**Comment by mike-morrison** [comment]
### Side Note => Documentation

Look at Aurelia's documentation.  It is beautiful.  It was done with typedoc, and the article style pages look like they are all coming from their .md files in their repo.  

If you look at the Aurelia organization, they have many libraries all following the same pattern.  

### Side Note / Dev Ops

Check it out though, they don't write their code in TypeScript.  They write the code in ES2016+.  They then use Facebook's [flow](http://flowtype.org/) (TypeScript is friendly to C# devs, where Flow is friendly to JS devs IMO) to take their ES6 with type annotations (compatible and designed to play nice with TypeScript), then generate the TypeScript definitions for them, which drives their docs.  Wow!

They are building a largely convention based framework with a bunch of really cool libraries (I am talking about their full dev stack and the pure libraries they are outputting).  They are using System.JS, and Babel with decorators and async/await to build all of their libraries.  It's a beautiful example of making something fun to code and outputting beautiful results.  Everything from docs to development experience.  

I am going to look at plugging into this when I can (some day), but with TypeScript instead of typed ES because I prefer code with as few _any's_ as possible and we need the C# devs.  Even though, it looks like Aurelia seems to prefer _any_ on the parts that don't matter for documentation.

### Bundling

Look at their [dist folder](https://github.com/aurelia/dependency-injection/tree/master/dist).  Multiple module types.  Take a peak at the commonjs library size.  17.5k.  I had to do a double take, it's not minimized yet!  It seems to cover most common scenarios.  

This is [the project](https://github.com/aurelia/framework) that brings in all the libraries and deps and builds the framework.

### TLDR;

Tying all of that together across that many separate projects/repos to build a framework and documentation while designing them to be used independently as well is amazing IMO.  

### I love gulp, but:

cards

**Comment by pbjorklund** [comment]
Very interesting, I will dig into Aurelias DI since I'm not 100% sold on Inversifys need to include <<<refs to get stuff building, it just don't feel production ready.

**Comment by pbjorklund** [comment]
I can see this not happening in the foreseeable future, a lot of other ground to cover before we can even think of introducting DI.
And with V1 fast approach I feel I missed the window to get something related to this in.

Closing this and will revisit it if it fits.
